### DISCUSSION TITLE
How to make the loading bar appear every time that I change my object source?

### DISCUSSION THREAD

**Discussion by gabrielgaldinomagalhaes** [google/model-viewer#4142] [Q&A] [answered]
I'm developing an application where the user can change the model that is being shown in the model-viewer, for exempla, they can change between the model1, model2, model3 and model4. I'm using as base the scene that I downloaded from the model-viewer/editor.
I'm using this one as base because I like the loading bar that appears in the center of the screen, better than the loading bar that appears in the top of the model-viewer window (the one that appears in the documentation).
Here are the different loading bar that I talked above:

This is the one that came with the scene that I downloaded from the model-viewer/editor (the one that I like):
![image](https://user-images.githubusercontent.com/45290323/221132563-66e692c9-60cc-45b2-9747-c261bc0b7688.png)

And this is the one that appears in the examples of the documentation (I don't want to use this, I want to use the first one):
![image](https://user-images.githubusercontent.com/45290323/221132940-242d992a-722e-496e-85eb-eb2f179ea5ae.png)

I'm having 2 issues with the loading bar, and needed help to fix this.
1- The first issue is, the loading bar appears when the user first load the page, when the first object is being loaded, but it doesn't appear when the user change the object for the model2, or the model 3. For example, it should appear when he change from the model 1 to the model 2 while the model 2 is still being loaded, so the user knows that something is still loading and the application is not crashed or stuck. Also when he changes to the model3 it should appear again, so the user knows that something is still loading and the application, the same thing. How can I make sure that the loading bar will appear in all model changings ??

This is the code that I'm using (the base that I downloaded in the model-viewer editor, the css is the default css too):

```
const modelViewer = document.querySelector("model-viewer");
	
// Handles loading the events for <model-viewer>'s slotted progress bar
const onProgress = (event) => {
	const progressBar = event.target.querySelector('.progress-bar');
	const updatingBar = event.target.querySelector('.update-bar');
	updatingBar.style.width = `${event.detail.totalProgress * 100}%`;
	if (event.detail.totalProgress === 1) {
		progressBar.classList.add('hide');
		event.target.removeEventListener('progress', onProgress);
	} else {
		progressBar.classList.remove('hide');
	} 
};
document.querySelector('model-viewer').addEventListener('progress', onProgress);
```

2- The second issue is: in the first time that the user reach the page, the loading bar appears but it completes (from 0 to 100%) before the model is loaded, and it keeps showing the poster instead. So the user thinks that the application is stuck or crashed, because he saw the loading bar, and the 3D model still didn't appear, it keeps showing the poster till the glb is loaded.
I wanted to make the loading bar runs while the poster is still being showed, and complete only after the model (.glb) is fully loaded. So he would be seeing the poster and the loading bar in the front of it. And the loading bar would disappear in the same time that the glb is loaded completely and the poster disappears.  Did you get that?

How can I fix this too?? Does someone know?

**Comment by elalish** [comment]
For 1, I think you just need to remove the `removeEventListener` line. For 2, that's strange - it should work how you expect. Perhaps you could share a screen recording? And perhaps check your JS console. I wonder if perhaps you're actually loading more models on page start than you think you are.

**Comment by gabrielgaldinomagalhaes** [comment, marked as answer]
It worked better like this:

```
// Handles loading the events for <model-viewer>'s slotted progress bar
const onProgress = (event) => {
  const progressBar = event.target.querySelector('.progress-bar');
  const updatingBar = event.target.querySelector('.update-bar');

  if (event.detail.totalProgress === 0) {
    progressBar.style.display = 'block';
    updatingBar.style.width = '0%';
  } else {
    updatingBar.style.width = `${event.detail.totalProgress * 100}%`;

    if (event.detail.totalProgress === 1) {
      setTimeout(() => {
        progressBar.style.display = 'none';
      }, 500);
    }
  }
};
document.querySelector('model-viewer').addEventListener('progress', onProgress);
```

**Reply by NeoAnthropocene** [reply]
This works better for sure.

Don't forget to update your div style. On the demo file it adds .hide, but with this one you should remove it from the div style. For instance:
```
<div class="progress-bar" slot="progress-bar">
          <div class="update-bar"></div>
</div>
```
